libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1820,6 +1846,8 @@ void TypeChecker::typeCheckFunctionCall(
 	if (_functionType->kind() == FunctionType::Kind::Declaration)
 	{
 		if (
+			m_currentContract &&
```

Comment:
Could this be an assertion instead?

Reply:
No, and this is tested in test/libsolidity/syntaxTests/freeFunctions/free_call_via_contract_type.sol:
```
contract C {
    function f() public pure {}
}
function fun() {
    C.f();
}
// ----
// TypeError 3419: (68-73): Cannot call function via contract type name.
```

Reply:
Ah, right - I actually meant the other one, one line below (as in: only functions within contract scopes should ever be available with type ``FunctionType::Kind::Declaration``, so ``_functionType->declaration().annotation().contract`` should never be ``nullptr`` here - that's why it could just be dereferenced before this PR, right?).
But also fine not to assert that.

Reply:
Ah this might be true! And it's important to realize this: References to free function can never be `::Declaration`. I'll add a test where we reference a free function via a module name to check if this is really the case.

Reply:
Ah, there already is a test and it does not violate the assertion: https://github.com/ethereum/solidity/pull/8834/commits/756aae0cccfae951e95db62d735c50ed9685116b#diff-9177bb8d3bc85509fb8289a4da7f6085

Reply:
> And it's important to realize this: References to free function can never be `::Declaration`.

Why is that?

Reply:
because Declaration always needs member access